grouper/models/public_key.py

```diff
@@ -18,3 +75,35 @@ class PublicKey(Model):
     public_key = Column(Text, nullable=False)
     fingerprint = Column(String(length=64), nullable=False, unique=True)
     created_on = Column(DateTime, default=datetime.utcnow, nullable=False)
+
+    def tags(self):
+        # type: () -> List[PublicKeyTag]
+        """Returns the list of tags that are assigned to this public key
+
+        Returns:
+            a list that contains all of the PublicKeyTags that are assigned to this public key
+        """
+        mappings = self.session.query(PublicKeyTagMap).filter_by(key_id=self.id).all()
+        return [mapping.tag for mapping in mappings]
+
+    def my_permissions(self):
+        # type: () -> List[Permission]
+        """Returns the permissions that this public key has. Namely, this the set of permissions
+        that the public key's owner has, intersected with the permissions allowed by this key's
+        tags
+
+        Returns:
+            a list of all permissions this public key has
+        """
+        my_perms = self.user.my_permissions()
+        for tag in self.tags():
+            my_perms = _permission_intersection(my_perms, tag.my_permissions())
+
+        return list(my_perms)
+
+    def pretty_my_permissions(self):
```

Comment:
We do this in a few different places. primarily in the `pemission_panel`. It'd be nice if this was kept in templating/display code vs something in business logic.